wasm: use wasmedge library soname in dlopen

hydai opened the pull request:
Notice: The WasmEdge 0.11.0 is planned to be released at the end of this month. Since the test will use the installer to install the corresponding version of WasmEdge, please merge this PR after the WasmEdge 0.11.0 release. Thanks.


Since wasmedge implemented library versioning in "e5af947d", use
the library's soname ("libwasmedge.so.0") instead of "libwasmedge_c.so"
to ensure crun can pin to a compatible ABI version.

Notice: the soname is shipped in wasmedge 0.11.0, before this release
the libwasmedge_c.so is unversioned.

For details, please refer to
[WasmEdge PR#1783](https://github.com/WasmEdge/WasmEdge/pull/1783)

Signed-off-by: hydai <hydai@secondstate.io>

hydai commented:
I set this to be `draft` first. The WasmEdge release on the fedora package also needs to update to 0.11.0 to provide the `libwasmedge.so`, `libwasmedge.so.0`, and `libwasmedge.so.0.0.0` in the `-devel` package.

flouthoc commented:
Closes: https://github.com/containers/crun/issues/990

hydai commented:
Hi @flouthoc 

The WasmEdge 0.11.0 is just released, I am going to update this PR and make the test pass.
I would also like to add an integration test to make sure the crun can execute a wasm image perfectly after the build is done. Does crun have examples to show how to add an integration test? So I can refer to them and make the wasm version one. Thanks.

hydai commented:
Also, I found that podman test failed, since this PR should not change anything related to the `podman` features, is that anything I can do for figuring out the reason?

flouthoc commented:
Hi @hydai , Following PR added test for `wasmedge build` https://github.com/containers/crun/pull/933, i think run_tests.sh can be extended to support some new tests or there are few examples here https://github.com/containers/crun/tree/main/tests



> Also, I found that podman test failed, since this PR should not change anything related to the `podman` features, is that anything I can do for figuring out the reason?

The failing podman test seems unrelated to crun, i think we might wanna skip that test in crun CI i'll open a PR for that after confirming.

hydai commented:
Thanks. I will update the test inside `run-test.sh` to extend the tests.

Another question is which docker image should I use. I found that the current image is using `debian:bullseye-slim`[1], but I found there are so many different base images used in the various tests. Is it fine to keep the current base image? Or do I need to use another one?

[1]: https://github.com/containers/crun/blob/main/tests/wasmedge-build/Dockerfile#L1

flouthoc commented:
Personally don't have a preference for base-image, looks okay to me if its already `debian:bullseye-slim` but feel free to change if you want to`. But I checked and different tests are using different base image so i guess there is no standard set. @giuseppe Could confirm if we want to stick to a single base image.

hydai commented:
To extend the test, I use the following instructions:

1. Provide an example rust application that prints a string via stdout.
2. Install and build the `hello.wasm` inside the Dockerfile. Because using `source $HOME/.cargo/env` will trigger an error by shellcheck.
3. Try to create a scratch container image with `config.json`, which is generated by `crun spec` and modified by adding the wasm-related annotations, and `container folder`, which contains a layer called `rootfs` and `hello.wasm`.
4. Use crun to execute the scratch container image.

The current status:
Maybe the formats or file structures of the container image are wrong, so the image cannot be executed correctly via `crun cun`. Will go back to reproduce in my environment.

The next step:
Since I am not able to create the image from scratch correctly, I plan to install and use `buildah` to generate the image first.

giuseppe commented:
@hydai could you please rebase?

It is fine to use a different base image, whatever fits better  :-)

hydai commented:
Sure thing, will process it tomorrow.

Also, I would like to change the based image to fedora, so I can get the newer version of podman easier for setting up the testing environment and running the integration tests. Thanks.

hydai commented:
Rebased.
The docker image is changed to `fedora:rawhide`. And I added a test to build the wasm image with `buildah` and execute it with `podman`. The test case is from https://github.com/containers/crun/blob/main/docs/wasm-wasi-example.md

This PR is ready to review. Thanks :-)

giuseppe reviewed (approved):
LGTM

@flouthoc PTAL and merge if you are fine with it

flouthoc reviewed (approved):
LGTM, THanks
